### DISCUSSION TITLE
Hitting (undocumented?) reusable workflow limit of 20

### DISCUSSION THREAD

**Discussion by JackCoplandGranta** [community/community#32192] [Actions]
I've just started hitting a reusable workflow limit of 20:
![image](https://user-images.githubusercontent.com/43538264/188811750-0167ac26-ae6b-439c-ae54-cdf69e6ee607.png)

I can't find any official docs of this limit, [I'd expect it to be mentioned here](https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations). Is this mentioned somewhere I've missed?

[I know the depth limit of reusable workflows was raised from 2 to 4 a bit ago](https://github.blog/changelog/2022-08-22-github-actions-improvements-to-reusable-workflows-2/), is there any hope of raising this limitation of them too?

**Comment by int128** [comment]
I tested 26 workflows in my public repository and it worked.
https://github.com/int128/sandbox/pull/178

~~However, I tested it in the organization private repository, the same error occurred.~~

Also I tested calling 26 different workflows, and got the error.
https://github.com/int128/sandbox/pull/179

```
too many workflows are referenced, limit is 20
```

**Comment by JackCoplandGranta** [comment]
This is still an ongoing issue for us, though the limitation is now at least documented. We're having to consider very tortured workarounds to deal with this.

**Comment by cruizba** [comment]
How it is possible that you try to apply "separation of concerns" and end up with more technical debt because of another undocumented limit? :(

**Comment by cruizba** [comment]
Did you find a workaround? @JackCoplandGranta

**Reply by JackCoplandGranta** [reply]
@cruizba There are some methods to achieve a similar effect, though none are as good as just a limit raise:
- Use composite actions (less capable feature, depth limit 10, not sure if there is a total use limit)
- Have a single reusable workflow with multiple jobs, and use job conditions (if:${{ inputs.job == 'job3' }}) to select which job to run using an input to the workflow. This 'counts' as a single reusable workflow, but can 'store' several jobs. If you have a complex job dependency structure, or lots of inputs, this may be hard to construct and maintain. 
- Use the [workflow_run trigger](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run), to chain together workflows. Workflow runs have a call depth limit of 3, but you can skirt this by manually triggering workflows via the rest API.
- Cut the Gordian knot and tactically duplicate code rather than re-use.

A combination of the above as most appropriate for your workflow structure will work, but obviously introduces various issues.

This limit is entirely artificial, you can construct workflows of arbitrary job and step complexity by abusing the features listed above and they won't have any issue starting or running.

**Reply by rjmcnamara** [reply]
> Have a single reusable workflow with multiple jobs, and use job conditions (if:${{ inputs.job == 'job3' }}) to select which job to run using an input to the workflow. This 'counts' as a single reusable workflow, but can 'store' several jobs. If you have a complex job dependency structure, or lots of inputs, this may be hard to construct and maintain.

@JackCoplandGranta have you tried that?  That was my same thought while trying to work around this arbitrary 20 workflow limit, but when I implemented the workaround, I found that workflows are included towards the limit whether or not they are actually called.

**Reply by JackCoplandGranta** [reply]
@rjmcnamara It sounds like you're still listing the reusable workflows and conditioning them at the top level. That's not what i meant. I mean combining several of your re-usable workflows into one, something like:

parent.yml, where you call the re-usable workflows:
```
name: "Full Chain"
on:
  workflow_dispatch:

jobs:

  Combined_Workflow:
    uses: ./.github/workflows/combined.yml
    secrets: inherit

```
combined.yml:
```
name: Combined Workflows

on:
  workflow_call:

jobs:

  job1:
    name: Job1
    runs-on: [self-hosted, windows, x64]
    steps:
      - run: sleep 5
      
  job2:
    name: Job2
    runs-on: [self-hosted, windows, x64]
    steps:
      - run: sleep 5
```

where Job1 and Job2 used to be re-usable workflows that you invoked from parent.yml as separate items. And if you need control of the jobs to run, pass down inputs.

**Comment by FelixSchladt** [comment]
I just had the same problem and I am even using self hosted runners.
How is this still a limitation in 2024???

**Reply by matthewcyy** [reply]
How is this still a limitation in 2025???

**Reply by MK-DBG** [reply]
How is this still a limitation in 2026???

**Comment by potiuk** [comment]
We've also hit the limit of it in our "code reuse" quest in Apache Airflow - we managed it by combining some jobs together, but it would be nice to have bigger limit.

**Comment by mhaley37** [comment]
This is an incredible frustrating limit to run up against -- both because it's unexpected and that it feels completely arbitrary.   Much like the arbitrary depth "call workflow" limit, there is no meaningful response from GitHub.  Why include such rich features for re-use of workflows only to introduce limits like this?   Open this work to community submissions and I'm sure you could find a contributor.

**Comment by chojnicki** [comment]
This is just very disappointing. Based on docs - reusable workflows supposed to help maintain bigger CI/CD, but because of limit, you should not use it on big CI/CD - awesome logic :P Someone in GH responsible for this functionality did not think this through. I'm waiting for change, but without big hopes.

**Comment by kishaningithub** [comment]
Hey Github,
     Show some love here! This one and call depth limit of 4 is such a huge limitation for managing big CI/CD

**Reply by kishaningithub** [reply]
Linking the depth limit thread here because these both needs to be addressed soon https://github.com/orgs/community/discussions/8488

**Comment by jasnova-mrj** [comment]
just hit 21. why!?

**Comment by tt-rkim** [comment]
This is so painful 😭 😭 😭 We have lots of buckets of tests for our software, but we also want developers to be able to run any set of pipelines they need to.

If we want to orchestrate more workflows from a central place, I'm not sure how to do that without further automation surrounding it. We would prefer to rely on inherent GitHub features instead.

**Comment by samudurand** [comment]
You can increase this limit if you have enterprise level support at the organisation level, you just need to make a request to support with the required number. We did it recently for a client and requested 50 as a limit, which was immediately granted. It might be something relatively recent because I haven't seen anyone else mentioning this in other platform.

**Reply by potiuk** [reply]
That's nice to know !

**Reply by nick-fields** [reply]
Interesting, we requested the same for our enterprise cloud account and received the following response:

> There is currently no way to bypass or increase this limit, however we've had similar requests from other customers so we've added your voice to the internal issue our engineering team are using to gauge customer interest. 
>  
> I can't say when or if this will be implemented, so please keep an eye on the [GitHub changelog](https://github.blog/changelog) page and the [GitHub public roadmap](https://github.com/github/roadmap) to stay up to date with new features and changes.

**Comment by paul-wade** [comment]
Call me crazy but documenting it so people know before they hit it might be cool? Or did I miss it? It's obviously not a new problem.

**Comment by ilmn-gcoffman** [comment]
I'm hitting the limit and I do not have 22 unique workflows being called.
I have a single workflow that calls at most 12 other reusable workflows.

**Reply by kishaningithub** [reply]
There could be transitive dependencies.

**Reply by ilmn-gcoffman** [reply]
According to the documentation of 3.15 GHE (the version we're using) the limit should be 50, not 20 regardless:
https://docs.github.com/en/enterprise-server@3.15/actions/reference/workflows-and-actions/reusing-workflow-configurations#limitations-of-reusable-workflows

**Reply by yeikel** [reply]
it seems that the limit for the latest GHE is still 20, it is 50 in GitHub.com 

https://github.blog/changelog/2025-11-06-new-releases-for-github-actions-november-2025/
